File: py/p290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, str: str) -> bool:
        '''Stolen from StefanPochmann posting.

        Very elegant but same runtime as my guy below.
        '''
        s = pattern
        t = str.split()
        return len(set(zip(s, t))) == len(set(s)) == len(
            set(t)) and len(s) == len(t)


class Solution2:
    def wordPattern(self, pattern: str, str: str) -> bool:
        '''My solution.

        Several wrong submissions because of poor problem definition.
        '''
        d = {}
        words = str.split(' ')

        # str.split will give [''] for empty string.
        if words[0] == '':
            if len(pattern) != 0:
                return False
        elif len(pattern) != len(words):
            return False

        for x in range(len(pattern)):
            c = pattern[x]
            word = words[x]
            if c in d:
                if d[c] != word:
                    return False
            else:
                d[c] = word

        # check to make sure dictionary words are unique.
        s = set()
        for val in d.values():
            if val in s:
                return False
            s.add(val)

        return True


if __name__ == '__main__':

    passed = True

    s = Solution()
    tests = (
        ('', '', True),
        ('', ' ', True),
        ('', ' grr', False),
        ('', 'dog cat cat dog', False),
        ('x', 'dog', True),
        ('x', 'dog cat cat dog', False),
        ('xy', 'dog cat', True),
        ('xy', 'dog cat cat dog', False),
        ('xx', 'dog dog', True),
        ('xx', 'dog dog cat cat', False),
        ('abba', 'dog cat cat dog', True),
        ('abba', 'dog cat cat fish', False),
        ('aaaa', 'dog cat cat dog', False),
        ('aaaa', 'dog dog dog dog', True),
        ('abba', 'dog dog dog dog', False),
        ('aaaa', 'dog cat cat dog', False),
        ('jquery', 'jquery', False),
    )

    for test in tests:
        pattern = test[0]
        words = test[1]
        exp = test[2]
        ans = s.wordPattern(pattern, words)
        if exp != ans:
            passed = False
            print(
                f'FAIL. pattern: {pattern}, words: {words}, exp: {exp}, ans: {ans}'
            )

File: leetcode/cpp/p290-word-pattern.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>

class Solution {
public:
    bool wordPattern(std::string pattern, std::string str)
    {
        std::unordered_map<char, std::string> d;
        size_t strIndex = 0;
        for (const char c : pattern) {
            while (str[strIndex] == ' ' && strIndex < str.size()) {
                ++strIndex;
            }
            if (strIndex == str.size()) {
                return false;
            }
            size_t wordEndIndex = strIndex;
            while (str[wordEndIndex] != ' ' && wordEndIndex < str.size()) {
                ++wordEndIndex;
            }
            const std::string word = str.substr(strIndex, wordEndIndex - strIndex);
            strIndex = wordEndIndex;
            const auto it = d.find(c);
            if (it == d.end()) {
                d.insert({ c, word });
            } else if (word != it->second) {
                return false;
            }
        }
        while (str[strIndex] == ' ' && strIndex < str.size()) {
            ++strIndex;
        }
        if (strIndex != str.size()) {
            return false;
        }
        // check to make sure dictionary words are unique.
        std::unordered_set<std::string> word_set;
        for (const auto& pair : d) {
            word_set.insert(pair.second);
        }
        return word_set.size() == d.size();
    }
};

int main(int argc, char* argv[])
{
    struct TestCase {
        std::string pattern;
        std::string s;
        bool exp;
    };
    const TestCase testCases[] = {
        { "", "", true },
        { "", " ", true },
        { "", " grr", false },
        { "", "dog cat cat dog", false },
        { "x", "dog", true },
        { "x", "dog cat cat dog", false },
        { "xy", "dog cat", true },
        { "xy", "dog cat cat dog", false },
        { "xx", "dog dog", true },
        { "xx", "dog dog cat cat", false },
        { "abba", "dog cat cat dog", true },
        { "abba", "dog cat cat fish", false },
        { "aaaa", "dog cat cat dog", false },
        { "aaaa", "dog dog dog dog", true },
        { "abba", "dog dog dog dog", false },
        { "aaaa", "dog cat cat dog", false },
        { "jquery", "jquery", false },
    };

    Solution s;
    for (const auto& tc : testCases) {
        const int ans = s.wordPattern(tc.pattern, tc.s);
        if (tc.exp != ans) {
            std::cout << "FAIL. pat: \"" << tc.pattern
                      << "\", str: \"" << tc.s
                      << "\", exp: " << tc.exp
                      << ", ans: " << ans << "\n";
        }
    }
    return 0;
}
